Copyright Jana Schaich Borg/Attribution-NonCommercial 4.0 International (CC BY-NC 4.0) 번역/보충설명: 황수민

# MySQL 연습문제 7: Inner Join으로 테이블 연결 하기

   
## 1. 2개의 테이블을 연결하는 Inner Join

In [17]:
%load_ext sql
%sql mysql://studentuser:studentpw@localhost/dognitiondb
%sql USE dognitiondb

The sql extension is already loaded. To reload it, use:
  %reload_ext sql
 * mysql://studentuser:***@localhost/dognitiondb
0 rows affected.


[]

In [14]:
%sql SELECT user_guid, dog_guid, rating \
FROM reviews \
GROUP BY user_guid \
LIMIT 100;

 * mysql://studentuser:***@localhost/dognitiondb
100 rows affected.


[(None, 'ce66c644-7144-11e5-ba71-058fbc01cf0b', 0),
 ('ce134492-7144-11e5-ba71-058fbc01cf0b', 'ce249a94-7144-11e5-ba71-058fbc01cf0b', 4),
 ('ce137868-7144-11e5-ba71-058fbc01cf0b', 'ce135cf2-7144-11e5-ba71-058fbc01cf0b', 3),
 ('ce137f20-7144-11e5-ba71-058fbc01cf0b', 'ce2212b0-7144-11e5-ba71-058fbc01cf0b', 2),
 ('ce137fca-7144-11e5-ba71-058fbc01cf0b', 'ce138a88-7144-11e5-ba71-058fbc01cf0b', 0),
 ('ce138268-7144-11e5-ba71-058fbc01cf0b', 'fd27dc52-7144-11e5-ba71-058fbc01cf0b', 0),
 ('ce1385c4-7144-11e5-ba71-058fbc01cf0b', 'ce136e36-7144-11e5-ba71-058fbc01cf0b', 6),
 ('ce1387cc-7144-11e5-ba71-058fbc01cf0b', 'ce1373ae-7144-11e5-ba71-058fbc01cf0b', None),
 ('ce2232cc-7144-11e5-ba71-058fbc01cf0b', 'ce78bf70-7144-11e5-ba71-058fbc01cf0b', 1),
 ('ce224208-7144-11e5-ba71-058fbc01cf0b', 'ce222a70-7144-11e5-ba71-058fbc01cf0b', 4),
 ('ce224cee-7144-11e5-ba71-058fbc01cf0b', 'ce224802-7144-11e5-ba71-058fbc01cf0b', 5),
 ('ce2261a2-7144-11e5-ba71-058fbc01cf0b', 'fd3fd514-7144-11e5-ba71-058fbc01cf0b', 0),
 ('ce226576-7144-11e5-ba71-058fbc01cf0b', 'ce2dc556-7144-11e5-ba71-058fbc01cf0b', 1),
 ('ce23ec8e-7144-11e5-ba71-058fbc01cf0b', 'ce2250b8-7144-11e5-ba71-058fbc01cf0b', 0),
 ('ce242e74-7144-11e5-ba71-058fbc01cf0b', 'ce24a16a-7144-11e5-ba71-058fbc01cf0b', 5),
 ('ce2434e6-7144-11e5-ba71-058fbc01cf0b', 'ce240cdc-7144-11e5-ba71-058fbc01cf0b', 2),
 ('ce24476a-7144-11e5-ba71-058fbc01cf0b', 'ce2422e4-7144-11e5-ba71-058fbc01cf0b', 0),
 ('ce244fd0-7144-11e5-ba71-058fbc01cf0b', 'ce82c7a4-7144-11e5-ba71-058fbc01cf0b', 0),
 ('ce2461dc-7144-11e5-ba71-058fbc01cf0b', 'ce243b6c-7144-11e5-ba71-058fbc01cf0b', 1),
 ('ce24a4ee-7144-11e5-ba71-058fbc01cf0b', 'ce246b82-7144-11e5-ba71-058fbc01cf0b', 6),
 ('ce24c50a-7144-11e5-ba71-058fbc01cf0b', 'ce24a3e0-7144-11e5-ba71-058fbc01cf0b', 0),
 ('ce24c8a2-7144-11e5-ba71-058fbc01cf0b', 'ce261fa4-7144-11e5-ba71-058fbc01cf0b', 0),
 ('ce24d130-7144-11e5-ba71-058fbc01cf0b', 'ce24b150-7144-11e5-ba71-058fbc01cf0b', 2),
 ('ce250e48-7144-11e5-ba71-058fbc01cf0b', 'ce24d7a2-7144-11e5-ba71-058fbc01cf0b', 0),
 ('ce2514c4-7144-11e5-ba71-058fbc01cf0b', 'ce24d96e-7144-11e5-ba71-058fbc01cf0b', None),
 ('ce251690-7144-11e5-ba71-058fbc01cf0b', 'ce284a04-7144-11e5-ba71-058fbc01cf0b', 0),
 ('ce251af0-7144-11e5-ba71-058fbc01cf0b', 'ce24e008-7144-11e5-ba71-058fbc01cf0b', 2),
 ('ce252c98-7144-11e5-ba71-058fbc01cf0b', 'ce3c2754-7144-11e5-ba71-058fbc01cf0b', 0),
 ('ce253cf6-7144-11e5-ba71-058fbc01cf0b', 'ce251d20-7144-11e5-ba71-058fbc01cf0b', 4),
 ('ce254534-7144-11e5-ba71-058fbc01cf0b', 'ce4596a4-7144-11e5-ba71-058fbc01cf0b', None),
 ('ce25482c-7144-11e5-ba71-058fbc01cf0b', 'ce252dba-7144-11e5-ba71-058fbc01cf0b', 6),
 ('ce2554ca-7144-11e5-ba71-058fbc01cf0b', 'ce253792-7144-11e5-ba71-058fbc01cf0b', None),
 ('ce257d2e-7144-11e5-ba71-058fbc01cf0b', 'ce254a02-7144-11e5-ba71-058fbc01cf0b', 8),
 ('ce2581a2-7144-11e5-ba71-058fbc01cf0b', 'ce254ab6-7144-11e5-ba71-058fbc01cf0b', None),
 ('ce258896-7144-11e5-ba71-058fbc01cf0b', 'ce255524-7144-11e5-ba71-058fbc01cf0b', 0),
 ('ce25a466-7144-11e5-ba71-058fbc01cf0b', 'ce80938a-7144-11e5-ba71-058fbc01cf0b', 0),
 ('ce25a6f0-7144-11e5-ba71-058fbc01cf0b', 'ce25894a-7144-11e5-ba71-058fbc01cf0b', 6),
 ('ce25b9d8-7144-11e5-ba71-058fbc01cf0b', 'ce3eea20-7144-11e5-ba71-058fbc01cf0b', 1),
 ('ce25c2ac-7144-11e5-ba71-058fbc01cf0b', 'ce25ade4-7144-11e5-ba71-058fbc01cf0b', 9),
 ('ce25f0c4-7144-11e5-ba71-058fbc01cf0b', 'ce25c0ea-7144-11e5-ba71-058fbc01cf0b', 2),
 ('ce25f772-7144-11e5-ba71-058fbc01cf0b', 'ce25cf22-7144-11e5-ba71-058fbc01cf0b', 4),
 ('ce25f9a2-7144-11e5-ba71-058fbc01cf0b', 'ce25edc2-7144-11e5-ba71-058fbc01cf0b', None),
 ('ce25fac4-7144-11e5-ba71-058fbc01cf0b', 'ce25eaa2-7144-11e5-ba71-058fbc01cf0b', 0),
 ('ce260168-7144-11e5-ba71-058fbc01cf0b', 'ce25f240-7144-11e5-ba71-058fbc01cf0b', 0),
 ('ce261aa4-7144-11e5-ba71-058fbc01cf0b', 'ce2609c4-7144-11e5-ba71-058fbc01cf0b', 0),
 ('ce261bd0-7144-11e5-ba71-058fbc01cf0b', 'ce260c1c-7144-11e5-ba71-058fbc01cf0b', 7),
 ('ce261edc-7144-11e5-ba71-058fbc01cf0b', 'ce2611b2-7144-11e

왜 rating을 평균값과 갯수(count)로 집계하는지 이해하지 못했으나 이는 한 강아지의 review는 여러개가 존재했기 때문이었다. 

In [23]:
%sql SELECT d.user_guid, d.dog_guid, d.breed_type, count(r.rating) AS count, AVG(r.rating) as AVG \
FROM dogs d, reviews r \
WHERE d.dog_guid=r.dog_guid AND d.user_guid=r.user_guid \
GROUP BY d.user_guid, d.dog_guid, d.breed_type \
HAVING count >= 10 \
ORDER BY AVG DESC \
LIMIT 200;


 * mysql://studentuser:***@localhost/dognitiondb
38 rows affected.


[('ce987914-7144-11e5-ba71-058fbc01cf0b', 'fdbf39f8-7144-11e5-ba71-058fbc01cf0b', 'Pure Breed', 12, Decimal('8.0000')),
 ('ce99bb12-7144-11e5-ba71-058fbc01cf0b', 'fdc09a82-7144-11e5-ba71-058fbc01cf0b', 'Popular Hybrid', 10, Decimal('5.0000')),
 ('ce984d2c-7144-11e5-ba71-058fbc01cf0b', 'fdbef330-7144-11e5-ba71-058fbc01cf0b', 'Pure Breed', 13, Decimal('4.5385')),
 ('ce99661c-7144-11e5-ba71-058fbc01cf0b', 'fdc0518a-7144-11e5-ba71-058fbc01cf0b', "Mixed Breed/ Other/ I Don't Know", 15, Decimal('3.7333')),
 ('ce473856-7144-11e5-ba71-058fbc01cf0b', 'fd684cf6-7144-11e5-ba71-058fbc01cf0b', 'Pure Breed', 18, Decimal('3.7222')),
 ('ce98b9b0-7144-11e5-ba71-058fbc01cf0b', 'fdbf66e4-7144-11e5-ba71-058fbc01cf0b', 'Pure Breed', 14, Decimal('3.5000')),
 ('ce99bb76-7144-11e5-ba71-058fbc01cf0b', 'fdc09b0e-7144-11e5-ba71-058fbc01cf0b', 'Popular Hybrid', 13, Decimal('3.2308')),
 ('ce9847dc-7144-11e5-ba71-058fbc01cf0b', 'fdbeedc2-7144-11e5-ba71-058fbc01cf0b', 'Pure Breed', 13, Decimal('3.1538')),
 ('ce9a7110-7144-11e5-ba71-058fbc01cf0b', 'fdc180be-7144-11e5-ba71-058fbc01cf0b', 'Cross Breed', 16, Decimal('3.0625')),
 ('ce6cd804-7144-11e5-ba71-058fbc01cf0b', 'fd6aabd6-7144-11e5-ba71-058fbc01cf0b', 'Pure Breed', 13, Decimal('3.0000')),
 ('ce98f3bc-7144-11e5-ba71-058fbc01cf0b', 'fdbfc0b2-7144-11e5-ba71-058fbc01cf0b', "Mixed Breed/ Other/ I Don't Know", 13, Decimal('2.6923')),
 ('ce982e46-7144-11e5-ba71-058fbc01cf0b', 'fdbedef4-7144-11e5-ba71-058fbc01cf0b', "Mixed Breed/ Other/ I Don't Know", 13, Decimal('2.4615')),
 ('ce980e3e-7144-11e5-ba71-058fbc01cf0b', 'fdbec8ce-7144-11e5-ba71-058fbc01cf0b', 'Pure Breed', 11, Decimal('2.4545')),
 ('ce9856f0-7144-11e5-ba71-058fbc01cf0b', 'fdbeff06-7144-11e5-ba71-058fbc01cf0b', 'Pure Breed', 10, Decimal('2.4000')),
 ('ce99049c-7144-11e5-ba71-058fbc01cf0b', 'fdbfe22c-7144-11e5-ba71-058fbc01cf0b', 'Pure Breed', 10, Decimal('2.4000')),
 ('ce981dc0-7144-11e5-ba71-058fbc01cf0b', 'fdbed7ce-7144-11e5-ba71-058fbc01cf0b', 'Pure Breed', 18, Decimal('2.3333')),
 ('ce9ac674-7144-11e5-ba71-058fbc01cf0b', 'fdc1c31c-7144-11e5-ba71-058fbc01cf0b', 'Pure Breed', 13, Decimal('2.3077')),
 ('ce99d674-7144-11e5-ba71-058fbc01cf0b', 'fdc0c49e-7144-11e5-ba71-058fbc01cf0b', "Mixed Breed/ Other/ I Don't Know", 18, Decimal('2.2778')),
 ('ce9aab30-7144-11e5-ba71-058fbc01cf0b', 'fdc19b30-7144-11e5-ba71-058fbc01cf0b', 'Pure Breed', 18, Decimal('2.2222')),
 ('ce982ed2-7144-11e5-ba71-058fbc01cf0b', 'fdbedf94-7144-11e5-ba71-058fbc01cf0b', "Mixed Breed/ Other/ I Don't Know", 11, Decimal('2.0000')),
 ('ce99d732-7144-11e5-ba71-058fbc01cf0b', 'fdc0c502-7144-11e5-ba71-058fbc01cf0b', 'Pure Breed', 15, Decimal('2.0000')),
 ('ce99049c-7144-11e5-ba71-058fbc01cf0b', 'fdbfddfe-7144-11e5-ba71-058fbc01cf0b', 'Pure Breed', 16, Decimal('1.9375')),
 ('ce7212a6-7144-11e5-ba71-058fbc01cf0b', 'fd7104ae-7144-11e5-ba71-058fbc01cf0b', "Mixed Breed/ Other/ I Don't Know", 10, Decimal('1.9000')),
 ('ce9a9500-7144-11e5-ba71-058fbc01cf0b', 'fdc198d8-7144-11e5-ba71-058fbc01cf0b', 'Pure Breed', 10, Decimal('1.8000')),
 ('ce97f926-7144-11e5-ba71-058fbc01cf0b', 'fdbe923c-7144-11e5-ba71-058fbc01cf0b', 'Pure Breed', 17, Decimal('1.7647')),
 ('ce6cf3fc-7144-11e5-ba71-058fbc01cf0b', 'fd6a7490-7144-11e5-ba71-058fbc01cf0b', 'Pure Breed', 20, Decimal('1.7500')),
 ('ce984f70-7144-11e5-ba71-058fbc01cf0b', 'fdbef5e2-7144-11e5-ba71-058fbc01cf0b', 'Pure Breed', 10, Decimal('1.7000')),
 ('ce98d6de-7144-11e5-ba71-058fbc01cf0b', 'fdbf9f10-7144-11e5-ba71-058fbc01cf0b', 'Cross Breed', 18, Decimal('1.6667')),
 ('ce9805ce-7144-11e5-ba71-058fbc01cf0b', 'fdbea952-7144-11e5-ba71-058fbc01cf0b', 'Popular Hybrid', 15, Decimal('1.6667')),
 ('ce98bca8-7144-11e5-ba71-058fbc01cf0b', 'fdbf686a-7144-11e5-ba71-058fbc01cf0b', "Mixed Breed/ Other/ I Don't Know", 10, Decimal('1.6000')),
 ('ce98b776-7144-11e5-ba71-058fbc01cf0b', 'fdbf6144-7144-11e5-ba71-058fbc01cf0b', 'Cross Breed', 11, Decimal('1.5455')),
 ('ce9ae4b0-7144-11e5-ba71-058fbc01cf0b', 'fdc1edec-7144-11e5-ba71-058fbc01cf0b', 'Pure Breed', 15, Decimal('1.

breed는 dog_guid와 동일 집계단위이므로 group by 문에서 실질적인 집계역할을 하지는 않으나, SELECT문에서 집계되지 않은 변수 명은 groupby의 절에 포함되어야한다는 원칙으로 인해 group by 뒤에 포함됨. 

In [9]:
%sql SELECT d.dog_guid AS DogID, d.user_guid AS UserID, AVG(r.rating) AS AvgRating, \
COUNT(r.rating) AS NumRatings, d.breed, d.breed_group, d.breed_type \
FROM dogs d, reviews r \
WHERE d.dog_guid=r.dog_guid AND d.user_guid=r.user_guid \
GROUP BY UserID, DogID, d.breed, d.breed_group, d.breed_type \
HAVING NumRatings >= 10 \
ORDER BY AvgRating DESC \
LIMIT 200;

 * mysql://studentuser:***@localhost/dognitiondb
38 rows affected.


[('fdbf39f8-7144-11e5-ba71-058fbc01cf0b', 'ce987914-7144-11e5-ba71-058fbc01cf0b', Decimal('8.0000'), 12, 'Canaan Dog', 'Herding', 'Pure Breed'),
 ('fdc09a82-7144-11e5-ba71-058fbc01cf0b', 'ce99bb12-7144-11e5-ba71-058fbc01cf0b', Decimal('5.0000'), 10, 'Golden Doodle', None, 'Popular Hybrid'),
 ('fdbef330-7144-11e5-ba71-058fbc01cf0b', 'ce984d2c-7144-11e5-ba71-058fbc01cf0b', Decimal('4.5385'), 13, 'Havanese', 'Toy', 'Pure Breed'),
 ('fdc0518a-7144-11e5-ba71-058fbc01cf0b', 'ce99661c-7144-11e5-ba71-058fbc01cf0b', Decimal('3.7333'), 15, 'Mixed', None, "Mixed Breed/ Other/ I Don't Know"),
 ('fd684cf6-7144-11e5-ba71-058fbc01cf0b', 'ce473856-7144-11e5-ba71-058fbc01cf0b', Decimal('3.7222'), 18, 'Labrador Retriever', 'Sporting', 'Pure Breed'),
 ('fdbf66e4-7144-11e5-ba71-058fbc01cf0b', 'ce98b9b0-7144-11e5-ba71-058fbc01cf0b', Decimal('3.5000'), 14, 'Golden Retriever', 'Sporting', 'Pure Breed'),
 ('fdc09b0e-7144-11e5-ba71-058fbc01cf0b', 'ce99bb76-7144-11e5-ba71-058fbc01cf0b', Decimal('3.2308'), 13, 'Golden Doodle', None, 'Popular Hybrid'),
 ('fdbeedc2-7144-11e5-ba71-058fbc01cf0b', 'ce9847dc-7144-11e5-ba71-058fbc01cf0b', Decimal('3.1538'), 13, 'Eurasier', None, 'Pure Breed'),
 ('fdc180be-7144-11e5-ba71-058fbc01cf0b', 'ce9a7110-7144-11e5-ba71-058fbc01cf0b', Decimal('3.0625'), 16, 'Chihuahua-Dachshund Mix', None, 'Cross Breed'),
 ('fd6aabd6-7144-11e5-ba71-058fbc01cf0b', 'ce6cd804-7144-11e5-ba71-058fbc01cf0b', Decimal('3.0000'), 13, 'American Pit Bull Terrier', None, 'Pure Breed'),
 ('fdbfc0b2-7144-11e5-ba71-058fbc01cf0b', 'ce98f3bc-7144-11e5-ba71-058fbc01cf0b', Decimal('2.6923'), 13, 'Mixed', None, "Mixed Breed/ Other/ I Don't Know"),
 ('fdbedef4-7144-11e5-ba71-058fbc01cf0b', 'ce982e46-7144-11e5-ba71-058fbc01cf0b', Decimal('2.4615'), 13, 'Mixed', None, "Mixed Breed/ Other/ I Don't Know"),
 ('fdbec8ce-7144-11e5-ba71-058fbc01cf0b', 'ce980e3e-7144-11e5-ba71-058fbc01cf0b', Decimal('2.4545'), 11, 'Basset Hound', 'Hound', 'Pure Breed'),
 ('fdbfe22c-7144-11e5-ba71-058fbc01cf0b', 'ce99049c-7144-11e5-ba71-058fbc01cf0b', Decimal('2.4000'), 10, 'Irish Water Spaniel', 'Sporting', 'Pure Breed'),
 ('fdbeff06-7144-11e5-ba71-058fbc01cf0b', 'ce9856f0-7144-11e5-ba71-058fbc01cf0b', Decimal('2.4000'), 10, 'Bearded Collie', 'Herding', 'Pure Breed'),
 ('fdbed7ce-7144-11e5-ba71-058fbc01cf0b', 'ce981dc0-7144-11e5-ba71-058fbc01cf0b', Decimal('2.3333'), 18, 'Havanese', 'Toy', 'Pure Breed'),
 ('fdc1c31c-7144-11e5-ba71-058fbc01cf0b', 'ce9ac674-7144-11e5-ba71-058fbc01cf0b', Decimal('2.3077'), 13, 'Labrador Retriever', 'Sporting', 'Pure Breed'),
 ('fdc0c49e-7144-11e5-ba71-058fbc01cf0b', 'ce99d674-7144-11e5-ba71-058fbc01cf0b', Decimal('2.2778'), 18, 'Other', None, "Mixed Breed/ Other/ I Don't Know"),
 ('fdc19b30-7144-11e5-ba71-058fbc01cf0b', 'ce9aab30-7144-11e5-ba71-058fbc01cf0b', Decimal('2.2222'), 18, 'Brittany', 'Sporting', 'Pure Breed'),
 ('fdbedf94-7144-11e5-ba71-058fbc01cf0b', 'ce982ed2-7144-11e5-ba71-058fbc01cf0b', Decimal('2.0000'), 11, 'Mixed', None, "Mixed Breed/ Other/ I Don't Know"),
 ('fdc0c502-7144-11e5-ba71-058fbc01cf0b', 'ce99d732-7144-11e5-ba71-058fbc01cf0b', Decimal('2.0000'), 15, 'Golden Retriever', 'Sporting', 'Pure Breed'),
 ('fdbfddfe-7144-11e5-ba71-058fbc01cf0b', 'ce99049c-7144-11e5-ba71-058fbc01cf0b', Decimal('1.9375'), 16, 'Irish Water Spaniel', 'Sporting', 'Pure Breed'),
 ('fd7104ae-7144-11e5-ba71-058fbc01cf0b', 'ce7212a6-7144-11e5-ba71-058fbc01cf0b', Decimal('1.9000'), 10, 'Mixed', None, "Mixed Breed/ Other/ I Don't Know"),
 ('fdc198d8-7144-11e5-ba71-058fbc01cf0b', 'ce9a9500-7144-11e5-ba71-058fbc01cf0b', Decimal('1.8000'), 10, 'German Shepherd Dog', 'Herding', 'Pure Breed'),
 ('fdbe923c-7144-11e5-ba71-058fbc01cf0b', 'ce97f926-7144-11e5-ba71-058fbc01cf0b', Decimal('1.7647'), 17, 'Beagle', 'Hound', 'Pure Breed'),
 ('fd6a7490-7144-11e5-ba71-058fbc01cf0b', 'ce6cf3fc-7144-11e5-ba71-058fbc01cf0b', Decimal('1.7500'), 20, 'Golden Retriever', 'Sporting', 'Pure Breed'),
 ('fdbef5e2-7144-11e5-ba71-058fbc01cf0b', 'ce984f70-7144-11e5-ba71-058fbc01cf0b', Decimal

앞 쿼리에서는 200명까지 출력하라고 했는데도 38개의 행인 테이블이 출력되었음, 그 이유는 10개의 이상의 리뷰를 남긴 개 주인의 수가 38명이기 때문임. 위의 쿼리에서 HAVING과 LIMIT절을 없에면 395 행의 테이블이 나옴

In [24]:
%sql SELECT d.dog_guid AS DogID, d.user_guid AS UserID, AVG(r.rating) AS AvgRating, \
COUNT(r.rating) AS NumRatings, d.breed, d.breed_group, d.breed_type \
FROM dogs d, reviews r \
WHERE d.dog_guid=r.dog_guid AND d.user_guid=r.user_guid \
GROUP BY UserID, DogID, d.breed, d.breed_group, d.breed_type \
ORDER BY AvgRating DESC;

 * mysql://studentuser:***@localhost/dognitiondb
395 rows affected.


[('fdbf94de-7144-11e5-ba71-058fbc01cf0b', 'ce98cc34-7144-11e5-ba71-058fbc01cf0b', Decimal('9.0000'), 2, 'Pembroke Welsh Corgi', 'Herding', 'Pure Breed'),
 ('fdbf178e-7144-11e5-ba71-058fbc01cf0b', 'ce986546-7144-11e5-ba71-058fbc01cf0b', Decimal('9.0000'), 1, 'Airedale Terrier', 'Terrier', 'Pure Breed'),
 ('fdc057ac-7144-11e5-ba71-058fbc01cf0b', 'ce996b9e-7144-11e5-ba71-058fbc01cf0b', Decimal('9.0000'), 1, 'Mixed', None, "Mixed Breed/ Other/ I Don't Know"),
 ('fdc19554-7144-11e5-ba71-058fbc01cf0b', 'ce9a8a24-7144-11e5-ba71-058fbc01cf0b', Decimal('9.0000'), 1, 'Miniature Pinscher', 'Toy', 'Pure Breed'),
 ('fdc147ac-7144-11e5-ba71-058fbc01cf0b', 'ce9a545a-7144-11e5-ba71-058fbc01cf0b', Decimal('9.0000'), 1, 'Tibetan Terrier', 'Non-Sporting', 'Pure Breed'),
 ('fdc1c880-7144-11e5-ba71-058fbc01cf0b', 'ce96c768-7144-11e5-ba71-058fbc01cf0b', Decimal('9.0000'), 1, 'Maltese', 'Toy', 'Pure Breed'),
 ('fdc121dc-7144-11e5-ba71-058fbc01cf0b', 'ce9a310a-7144-11e5-ba71-058fbc01cf0b', Decimal('9.0000'), 1, 'Mixed', None, "Mixed Breed/ Other/ I Don't Know"),
 ('fdbedd14-7144-11e5-ba71-058fbc01cf0b', 'ce975994-7144-11e5-ba71-058fbc01cf0b', Decimal('9.0000'), 1, 'Mixed', None, "Mixed Breed/ Other/ I Don't Know"),
 ('fdc059fa-7144-11e5-ba71-058fbc01cf0b', 'ce996dc4-7144-11e5-ba71-058fbc01cf0b', Decimal('9.0000'), 1, 'Irish Water Spaniel', 'Sporting', 'Pure Breed'),
 ('fdc019c2-7144-11e5-ba71-058fbc01cf0b', 'ce99489e-7144-11e5-ba71-058fbc01cf0b', Decimal('9.0000'), 1, 'Mixed', None, "Mixed Breed/ Other/ I Don't Know"),
 ('fd41bca8-7144-11e5-ba71-058fbc01cf0b', 'ce7b055a-7144-11e5-ba71-058fbc01cf0b', Decimal('8.0000'), 1, 'Labrador Retriever', 'Sporting', 'Pure Breed'),
 ('fdbf39f8-7144-11e5-ba71-058fbc01cf0b', 'ce987914-7144-11e5-ba71-058fbc01cf0b', Decimal('8.0000'), 12, 'Canaan Dog', 'Herding', 'Pure Breed'),
 ('fdc1e1a8-7144-11e5-ba71-058fbc01cf0b', 'ce9ad934-7144-11e5-ba71-058fbc01cf0b', Decimal('7.5000'), 2, 'Mixed', None, "Mixed Breed/ Other/ I Don't Know"),
 ('fdc2417a-7144-11e5-ba71-058fbc01cf0b', 'ce9b35e6-7144-11e5-ba71-058fbc01cf0b', Decimal('7.0000'), 1, 'Bichon Frise', 'Non-Sporting', 'Pure Breed'),
 ('fdc1baca-7144-11e5-ba71-058fbc01cf0b', 'ce9ac2fa-7144-11e5-ba71-058fbc01cf0b', Decimal('7.0000'), 1, 'Poodle-Miniature Schnauzer Mix', None, 'Cross Breed'),
 ('fdbe7a40-7144-11e5-ba71-058fbc01cf0b', 'ce97df9a-7144-11e5-ba71-058fbc01cf0b', Decimal('7.0000'), 6, 'Mixed', None, "Mixed Breed/ Other/ I Don't Know"),
 ('fdbe8774-7144-11e5-ba71-058fbc01cf0b', 'ce97ebfc-7144-11e5-ba71-058fbc01cf0b', Decimal('7.0000'), 4, 'Labrador Retriever', 'Sporting', 'Pure Breed'),
 ('fdbf940c-7144-11e5-ba71-058fbc01cf0b', 'ce98cb80-7144-11e5-ba71-058fbc01cf0b', Decimal('7.0000'), 1, 'Soft Coated Wheaten Terrier', 'Terrier', 'Pure Breed'),
 ('fdc1706a-7144-11e5-ba71-058fbc01cf0b', 'ce9a6846-7144-11e5-ba71-058fbc01cf0b', Decimal('6.8000'), 5, 'Mixed', None, "Mixed Breed/ Other/ I Don't Know"),
 ('fdbf8598-7144-11e5-ba71-058fbc01cf0b', 'ce98c46e-7144-11e5-ba71-058fbc01cf0b', Decimal('6.5000'), 2, 'Labrador Retriever', 'Sporting', 'Pure Breed'),
 ('fdc0c2aa-7144-11e5-ba71-058fbc01cf0b', 'ce99d322-7144-11e5-ba71-058fbc01cf0b', Decimal('6.5000'), 2, 'Border Collie', 'Herding', 'Pure Breed'),
 ('fdc0827c-7144-11e5-ba71-058fbc01cf0b', 'ce99af32-7144-11e5-ba71-058fbc01cf0b', Decimal('6.0000'), 1, 'Irish Setter', 'Sporting', 'Pure Breed'),
 ('fd442826-7144-11e5-ba71-058fbc01cf0b', 'ce275612-7144-11e5-ba71-058fbc01cf0b', Decimal('6.0000'), 1, 'Mixed', None, "Mixed Breed/ Other/ I Don't Know"),
 ('fdc1ab52-7144-11e5-ba71-058fbc01cf0b', 'ce9ab8f0-7144-11e5-ba71-058fbc01cf0b', Decimal('6.0000'), 1, 'Golden Retriever', 'Sporting', 'Pure Breed'),
 ('fdc0d75e-7144-11e5-ba71-058fbc01cf0b', 'ce99e466-7144-11e5-ba71-058fbc01cf0b', Decimal('6.0000'), 1, 'Dachshund', 'Hound', 'Pure Breed'),
 ('fdc01f76-7144-11e5-ba71-058fbc01cf0b', 'ce994f1a-7144-11e5-ba71-058fbc01cf0b', Decimal('6.0000'), 1, 'Golden Doodle', None, 'Popular Hybrid'),
 ('fdc0345c-7144-11e5-ba71-058fbc01cf0b', 'ce99

위의 쿼리 결과를 통해 (A) 개 주인들은 리뷰들을 많이 남기지 않았다 (B) 리뷰를 남긴 결과를 살펴보면 결과에 그렇게 놀라지 않았다 는 것을 알수 있음   

**질문 1: reviews테이블에는 몇개의 고유 dog_guid가 있나요?**  
**질문 2: dogs테이블에는 몇개의 고유 dog_guid가 있나요?**  
**질문 3: reviews테이블에는 몇개의 고유 user_guid가 있나요?**  
**질문 4: dogs테이블에는 몇개의 고유 user_guid가 있나요?**  

In [25]:
#question 1 
%sql SELECT COUNT(DISTINCT(dog_guid)) AS count FROM reviews

 * mysql://studentuser:***@localhost/dognitiondb
1 rows affected.


[(5991,)]

In [26]:
#question 2 
%sql SELECT COUNT(DISTINCT(dog_guid)) AS count FROM dogs;

 * mysql://studentuser:***@localhost/dognitiondb
1 rows affected.


[(35050,)]

In [27]:
#question 3
%sql SELECT COUNT(DISTINCT(user_guid)) AS count FROM reviews;

 * mysql://studentuser:***@localhost/dognitiondb
1 rows affected.


[(5586,)]

In [29]:
#question 3
%sql SELECT COUNT(DISTINCT(user_guid)) AS count FROM dogs;

 * mysql://studentuser:***@localhost/dognitiondb
1 rows affected.


[(30967,)]

위의 쿼리 결과를 보면 다음과 같은 상황들을 알 수 있습니다:

+ reviews와 dogs테이블에 있는 고객들은 여러마리의 개가 있습니다
+ dogs테이블에 reviews테이블보다 훨씬 많은 고유의 dog_guids (개)과 user_guids (사용자)가 있습니다
+ reviews테이블에 INNER JOIN으로 연결된 테이블보다 훨씬 많은 dog_guids와 user_guids가 있습니다

또 테스트 해야 하는 부분이 있습니다.  
   
**밑에 있는 첫번째 셀에는 INNER JOIN의 쿼리를 실행하되 WHERE절에 dog_guid만 명시하고, 두번째 셀에는 user_guid만 명시해보세요:**

In [31]:
%sql SELECT d.dog_guid AS DogID, d.user_guid AS UserID, AVG(r.rating) AS AvgRating, \
COUNT(r.rating) AS NumRatings, d.breed, d.breed_group, d.breed_type \
FROM dogs d, reviews r \
WHERE d.dog_guid=r.dog_guid \
GROUP BY UserID, DogID, d.breed, d.breed_group, d.breed_type \
ORDER BY AvgRating DESC;

 * mysql://studentuser:***@localhost/dognitiondb
395 rows affected.


[('fdbf94de-7144-11e5-ba71-058fbc01cf0b', 'ce98cc34-7144-11e5-ba71-058fbc01cf0b', Decimal('9.0000'), 2, 'Pembroke Welsh Corgi', 'Herding', 'Pure Breed'),
 ('fdbf178e-7144-11e5-ba71-058fbc01cf0b', 'ce986546-7144-11e5-ba71-058fbc01cf0b', Decimal('9.0000'), 1, 'Airedale Terrier', 'Terrier', 'Pure Breed'),
 ('fdc057ac-7144-11e5-ba71-058fbc01cf0b', 'ce996b9e-7144-11e5-ba71-058fbc01cf0b', Decimal('9.0000'), 1, 'Mixed', None, "Mixed Breed/ Other/ I Don't Know"),
 ('fdc19554-7144-11e5-ba71-058fbc01cf0b', 'ce9a8a24-7144-11e5-ba71-058fbc01cf0b', Decimal('9.0000'), 1, 'Miniature Pinscher', 'Toy', 'Pure Breed'),
 ('fdc147ac-7144-11e5-ba71-058fbc01cf0b', 'ce9a545a-7144-11e5-ba71-058fbc01cf0b', Decimal('9.0000'), 1, 'Tibetan Terrier', 'Non-Sporting', 'Pure Breed'),
 ('fdc1c880-7144-11e5-ba71-058fbc01cf0b', 'ce96c768-7144-11e5-ba71-058fbc01cf0b', Decimal('9.0000'), 1, 'Maltese', 'Toy', 'Pure Breed'),
 ('fdc121dc-7144-11e5-ba71-058fbc01cf0b', 'ce9a310a-7144-11e5-ba71-058fbc01cf0b', Decimal('9.0000'), 1, 'Mixed', None, "Mixed Breed/ Other/ I Don't Know"),
 ('fdbedd14-7144-11e5-ba71-058fbc01cf0b', 'ce975994-7144-11e5-ba71-058fbc01cf0b', Decimal('9.0000'), 1, 'Mixed', None, "Mixed Breed/ Other/ I Don't Know"),
 ('fdc059fa-7144-11e5-ba71-058fbc01cf0b', 'ce996dc4-7144-11e5-ba71-058fbc01cf0b', Decimal('9.0000'), 1, 'Irish Water Spaniel', 'Sporting', 'Pure Breed'),
 ('fdc019c2-7144-11e5-ba71-058fbc01cf0b', 'ce99489e-7144-11e5-ba71-058fbc01cf0b', Decimal('9.0000'), 1, 'Mixed', None, "Mixed Breed/ Other/ I Don't Know"),
 ('fdbf39f8-7144-11e5-ba71-058fbc01cf0b', 'ce987914-7144-11e5-ba71-058fbc01cf0b', Decimal('8.0000'), 12, 'Canaan Dog', 'Herding', 'Pure Breed'),
 ('fd41bca8-7144-11e5-ba71-058fbc01cf0b', 'ce7b055a-7144-11e5-ba71-058fbc01cf0b', Decimal('8.0000'), 1, 'Labrador Retriever', 'Sporting', 'Pure Breed'),
 ('fdc1e1a8-7144-11e5-ba71-058fbc01cf0b', 'ce9ad934-7144-11e5-ba71-058fbc01cf0b', Decimal('7.5000'), 2, 'Mixed', None, "Mixed Breed/ Other/ I Don't Know"),
 ('fdc2417a-7144-11e5-ba71-058fbc01cf0b', 'ce9b35e6-7144-11e5-ba71-058fbc01cf0b', Decimal('7.0000'), 1, 'Bichon Frise', 'Non-Sporting', 'Pure Breed'),
 ('fdc1baca-7144-11e5-ba71-058fbc01cf0b', 'ce9ac2fa-7144-11e5-ba71-058fbc01cf0b', Decimal('7.0000'), 1, 'Poodle-Miniature Schnauzer Mix', None, 'Cross Breed'),
 ('fdbe7a40-7144-11e5-ba71-058fbc01cf0b', 'ce97df9a-7144-11e5-ba71-058fbc01cf0b', Decimal('7.0000'), 6, 'Mixed', None, "Mixed Breed/ Other/ I Don't Know"),
 ('fdbf940c-7144-11e5-ba71-058fbc01cf0b', 'ce98cb80-7144-11e5-ba71-058fbc01cf0b', Decimal('7.0000'), 1, 'Soft Coated Wheaten Terrier', 'Terrier', 'Pure Breed'),
 ('fdbe8774-7144-11e5-ba71-058fbc01cf0b', 'ce97ebfc-7144-11e5-ba71-058fbc01cf0b', Decimal('7.0000'), 4, 'Labrador Retriever', 'Sporting', 'Pure Breed'),
 ('fdc1706a-7144-11e5-ba71-058fbc01cf0b', 'ce9a6846-7144-11e5-ba71-058fbc01cf0b', Decimal('6.8000'), 5, 'Mixed', None, "Mixed Breed/ Other/ I Don't Know"),
 ('fdbf8598-7144-11e5-ba71-058fbc01cf0b', 'ce98c46e-7144-11e5-ba71-058fbc01cf0b', Decimal('6.5000'), 2, 'Labrador Retriever', 'Sporting', 'Pure Breed'),
 ('fdc0c2aa-7144-11e5-ba71-058fbc01cf0b', 'ce99d322-7144-11e5-ba71-058fbc01cf0b', Decimal('6.5000'), 2, 'Border Collie', 'Herding', 'Pure Breed'),
 ('fdc0827c-7144-11e5-ba71-058fbc01cf0b', 'ce99af32-7144-11e5-ba71-058fbc01cf0b', Decimal('6.0000'), 1, 'Irish Setter', 'Sporting', 'Pure Breed'),
 ('fdc1ab52-7144-11e5-ba71-058fbc01cf0b', 'ce9ab8f0-7144-11e5-ba71-058fbc01cf0b', Decimal('6.0000'), 1, 'Golden Retriever', 'Sporting', 'Pure Breed'),
 ('fdc0d75e-7144-11e5-ba71-058fbc01cf0b', 'ce99e466-7144-11e5-ba71-058fbc01cf0b', Decimal('6.0000'), 1, 'Dachshund', 'Hound', 'Pure Breed'),
 ('fdc01f76-7144-11e5-ba71-058fbc01cf0b', 'ce994f1a-7144-11e5-ba71-058fbc01cf0b', Decimal('6.0000'), 1, 'Golden Doodle', None, 'Popular Hybrid'),
 ('fd442826-7144-11e5-ba71-058fbc01cf0b', 'ce275612-7144-11e5-ba71-058fbc01cf0b', Decimal('6.0000'), 1, 'Mixed', None, "Mixed Breed/ Other/ I Don't Know"),
 ('fdc0345c-7144-11e5-ba71-058fbc01cf0b', 'ce99

In [30]:
%sql SELECT d.dog_guid AS DogID, d.user_guid AS UserID, AVG(r.rating) AS AvgRating, \
COUNT(r.rating) AS NumRatings, d.breed, d.breed_group, d.breed_type \
FROM dogs d, reviews r \
WHERE d.user_guid=r.user_guid \
GROUP BY UserID, DogID, d.breed, d.breed_group, d.breed_type \
ORDER BY AvgRating DESC;

 * mysql://studentuser:***@localhost/dognitiondb
6267 rows affected.


[('fd423f0c-7144-11e5-ba71-058fbc01cf0b', 'ce25c2ac-7144-11e5-ba71-058fbc01cf0b', Decimal('9.0000'), 2, 'Wire Fox Terrier', 'Terrier', 'Pure Breed'),
 ('fd42ec90-7144-11e5-ba71-058fbc01cf0b', 'ce264330-7144-11e5-ba71-058fbc01cf0b', Decimal('9.0000'), 3, 'Mastiff', 'Working', 'Pure Breed'),
 ('fd42ed6c-7144-11e5-ba71-058fbc01cf0b', 'ce264330-7144-11e5-ba71-058fbc01cf0b', Decimal('9.0000'), 3, 'French Bulldog', 'Non-Sporting', 'Pure Breed'),
 ('fd43c4b2-7144-11e5-ba71-058fbc01cf0b', 'ce27023e-7144-11e5-ba71-058fbc01cf0b', Decimal('9.0000'), 1, 'Mixed', '', "Mixed Breed/ Other/ I Don't Know"),
 ('fd43c516-7144-11e5-ba71-058fbc01cf0b', 'ce27023e-7144-11e5-ba71-058fbc01cf0b', Decimal('9.0000'), 1, 'Mixed', None, "Mixed Breed/ Other/ I Don't Know"),
 ('fd51c72e-7144-11e5-ba71-058fbc01cf0b', 'ce32fd00-7144-11e5-ba71-058fbc01cf0b', Decimal('9.0000'), 1, 'Miniature Schnauzer', 'Terrier', 'Pure Breed'),
 ('fd66d024-7144-11e5-ba71-058fbc01cf0b', 'ce40e3f2-7144-11e5-ba71-058fbc01cf0b', Decimal('9.0000'), 1, 'Brittany', 'Sporting', 'Pure Breed'),
 ('fd67ae40-7144-11e5-ba71-058fbc01cf0b', 'ce458498-7144-11e5-ba71-058fbc01cf0b', Decimal('9.0000'), 1, 'Mixed', '', "Mixed Breed/ Other/ I Don't Know"),
 ('fd67aea4-7144-11e5-ba71-058fbc01cf0b', 'ce458498-7144-11e5-ba71-058fbc01cf0b', Decimal('9.0000'), 1, 'Rat Terrier', '', 'Pure Breed'),
 ('fd67b71e-7144-11e5-ba71-058fbc01cf0b', 'ce45974e-7144-11e5-ba71-058fbc01cf0b', Decimal('9.0000'), 1, 'Schipperke', 'Non-Sporting', 'Pure Breed'),
 ('fd67b782-7144-11e5-ba71-058fbc01cf0b', 'ce45974e-7144-11e5-ba71-058fbc01cf0b', Decimal('9.0000'), 1, 'Mixed', None, "Mixed Breed/ Other/ I Don't Know"),
 ('fd684e2c-7144-11e5-ba71-058fbc01cf0b', 'ce4739a0-7144-11e5-ba71-058fbc01cf0b', Decimal('9.0000'), 1, 'Labrador Retriever-Golden Retriever Mix', None, 'Cross Breed'),
 ('fd68df90-7144-11e5-ba71-058fbc01cf0b', 'ce66539e-7144-11e5-ba71-058fbc01cf0b', Decimal('9.0000'), 1, 'Beagle-Rat Terrier Mix', None, 'Cross Breed'),
 ('fd6ac47c-7144-11e5-ba71-058fbc01cf0b', 'ce6d4c80-7144-11e5-ba71-058fbc01cf0b', Decimal('9.0000'), 1, 'German Shepherd Dog-Boxer Mix', None, 'Cross Breed'),
 ('fd6c9ba8-7144-11e5-ba71-058fbc01cf0b', 'ce6e58a0-7144-11e5-ba71-058fbc01cf0b', Decimal('9.0000'), 1, 'Beagle-Shetland Sheepdog Mix', None, 'Cross Breed'),
 ('fd6cf774-7144-11e5-ba71-058fbc01cf0b', 'ce6e7c86-7144-11e5-ba71-058fbc01cf0b', Decimal('9.0000'), 1, 'Labrador Retriever', 'Sporting', 'Pure Breed'),
 ('fd6d271c-7144-11e5-ba71-058fbc01cf0b', 'ce6ec4c0-7144-11e5-ba71-058fbc01cf0b', Decimal('9.0000'), 1, 'Australian Shepherd', 'Herding', 'Pure Breed'),
 ('fd6d535e-7144-11e5-ba71-058fbc01cf0b', 'ce6ecd80-7144-11e5-ba71-058fbc01cf0b', Decimal('9.0000'), 1, 'Vizsla', 'Sporting', 'Pure Breed'),
 ('fd6dcdca-7144-11e5-ba71-058fbc01cf0b', 'ce6f3dce-7144-11e5-ba71-058fbc01cf0b', Decimal('9.0000'), 1, 'Mixed', None, "Mixed Breed/ Other/ I Don't Know"),
 ('fd6dd1e4-7144-11e5-ba71-058fbc01cf0b', 'ce6f4116-7144-11e5-ba71-058fbc01cf0b', Decimal('9.0000'), 1, 'Mixed', None, "Mixed Breed/ Other/ I Don't Know"),
 ('fd6dfdd6-7144-11e5-ba71-058fbc01cf0b', 'ce6f6754-7144-11e5-ba71-058fbc01cf0b', Decimal('9.0000'), 1, 'Mixed', None, "Mixed Breed/ Other/ I Don't Know"),
 ('fd6e4ef8-7144-11e5-ba71-058fbc01cf0b', 'ce6fb3a8-7144-11e5-ba71-058fbc01cf0b', Decimal('9.0000'), 1, 'German Shepherd Dog', 'Herding', 'Pure Breed'),
 ('fd6e8076-7144-11e5-ba71-058fbc01cf0b', 'ce6fcdb6-7144-11e5-ba71-058fbc01cf0b', Decimal('9.0000'), 1, 'Labrador Retriever', 'Sporting', 'Pure Breed'),
 ('fd6ee764-7144-11e5-ba71-058fbc01cf0b', 'ce6fe8c8-7144-11e5-ba71-058fbc01cf0b', Decimal('9.0000'), 1, 'Pyrenean Shepherd', 'Herding', 'Pure Breed'),
 ('fd6ecfcc-7144-11e5-ba71-058fbc01cf0b', 'ce7024aa-7144-11e5-ba71-058fbc01cf0b', Decimal('9.0000'), 1, 'Mixed', None, "Mixed Breed/ Other/ I Don't Know"),
 ('fd6eeb56-7144-11e5-ba71-058fbc01cf0b', 'ce70353a-7144-11e5-ba71-058fbc01cf0b', Decimal('9.0000'), 3, 'Australian Shepherd', 'Herding', 'Pure Breed'),
 ('fd6ef39e-7144-11e5-ba71-058fbc0

dog_guid로만 테이블을 연결하면 395개의 행의 답이 출력되는 반면에 user_guid로만 테이블을 연결하면 6267개의 행의 답이 출력됨. 이 뜻은: 

+ reviews테이블에 있는 user_guid들은 모두 다 dogs 테이블에 있음
+ 6000개가 넘은 dog_guids가 reviews테이블에 있지만, 그중에서 395개의 dog_guids만 dogs테이블에 있음

**질문 5: "Yawn Warm-up" 게임을 완료한 개들의 user_guid, dog_guid, breed, breed_type과 breed_group을 출력하세요 (WHERE절에서 dog_guids로만 연결하면 20,845행의 답을 얻게 됩니다. 여기서 힌트는 완료된 테스트의 데이터가 어느 테이블에 있는지 알아내는 것입니다. dogs테이블도, reviews테이블도 아닙니다)?**

In [11]:
%sql SELECT DISTINCT(test_name) FROM complete_tests;

 * mysql://studentuser:***@localhost/dognitiondb
40 rows affected.


[('Yawn Warm-up',),
 ('Yawn Game',),
 ('Eye Contact Warm-up',),
 ('Eye Contact Game',),
 ('Treat Warm-up',),
 ('Arm Pointing',),
 ('Foot Pointing',),
 ('Watching',),
 ('Turn Your Back',),
 ('Cover Your Eyes',),
 ('Watching - Part 2',),
 ('One Cup Warm-up',),
 ('Two Cup Warm-up',),
 ('Memory versus Pointing',),
 ('Memory versus Smell',),
 ('Delayed Cup Game',),
 ('Inferential Reasoning Warm-up',),
 ('Inferential Reasoning Game',),
 ('Physical Reasoning Warm-up',),
 ('Physical Reasoning Game',),
 ('Navigation Warm-up',),
 ('Navigation Learning',),
 ('Navigation Game',),
 ('Impossible Task Warm-up',),
 ('Impossible Task Game',),
 ('Numerosity Warm-Up',),
 ('5 vs 1 Game',),
 ('3 vs 1 Game',),
 ('Warm-Up',),
 ('1 vs 1 Game',),
 ('Smell Game',),
 ('Switch',),
 ('Slide',),
 ('Self Control Game',),
 ('Expression Game',),
 ('Different Perspective',),
 ('Shared Perspective',),
 ('Stair Game',),
 ('Shaker Warm-Up',),
 ('Shaker Game',)]

In [ ]:
%sql SELECT d.user_guid, d.dog_guid, d.breed, d.breed_type, c.test_name FROM dogs d,complete_tests c \
WHERE c.test_name = "Yawn Warm-up" AND d.dog_guid=c.dog_guid; 

 * mysql://studentuser:***@localhost/dognitiondb


In [18]:
%sql SELECT c.user_guid AS UserID, u.state, u.zip, d.dog_guid AS DogID, d.breed, d.breed_type, d.breed_group \
FROM dogs d, complete_tests c, users u \
WHERE d.dog_guid=c.dog_guid \
   AND c.user_guid=u.user_guid \
   AND c.test_name="Yawn Warm-up"

 * mysql://studentuser:***@localhost/dognitiondb
0 rows affected.


[]

## 2. 두개 이상의 테이블을 연결하기

**질문 6: 적어도 하나의 Dognition 테스트를 완료한 골든 리트리버 (Golden Retrievers)의 user_guid, membership_type과 dog_guid를 출력하세요 (답은 711개 행 입니다)?**

In [35]:
%sql SELECT * FROM complete_tests LIMIT 10;

 * mysql://studentuser:***@localhost/dognitiondb
10 rows affected.


[(datetime.datetime(2013, 2, 5, 18, 26, 54), datetime.datetime(2013, 2, 5, 18, 26, 54), None, 'fd27b86c-7144-11e5-ba71-058fbc01cf0b', 'Yawn Warm-up', 'Empathy'),
 (datetime.datetime(2013, 2, 5, 18, 31, 3), datetime.datetime(2013, 2, 5, 18, 31, 3), None, 'fd27b86c-7144-11e5-ba71-058fbc01cf0b', 'Yawn Game', 'Empathy'),
 (datetime.datetime(2013, 2, 5, 18, 32, 4), datetime.datetime(2013, 2, 5, 18, 32, 4), None, 'fd27b86c-7144-11e5-ba71-058fbc01cf0b', 'Eye Contact Warm-up', 'Empathy'),
 (datetime.datetime(2013, 2, 5, 18, 32, 25), datetime.datetime(2013, 2, 5, 18, 32, 25), None, 'fd27b86c-7144-11e5-ba71-058fbc01cf0b', 'Eye Contact Game', 'Empathy'),
 (datetime.datetime(2013, 2, 5, 18, 32, 56), datetime.datetime(2013, 2, 5, 18, 32, 56), None, 'fd27b86c-7144-11e5-ba71-058fbc01cf0b', 'Treat Warm-up', 'Communication'),
 (datetime.datetime(2013, 2, 5, 18, 33, 15), datetime.datetime(2013, 2, 5, 18, 33, 15), None, 'fd27b86c-7144-11e5-ba71-058fbc01cf0b', 'Arm Pointing', 'Communication'),
 (datetime.datetime(2013, 2, 5, 18, 33, 33), datetime.datetime(2013, 2, 5, 18, 33, 33), None, 'fd27b86c-7144-11e5-ba71-058fbc01cf0b', 'Foot Pointing', 'Communication'),
 (datetime.datetime(2013, 2, 5, 18, 33, 59), datetime.datetime(2013, 2, 5, 18, 33, 59), None, 'fd27b86c-7144-11e5-ba71-058fbc01cf0b', 'Watching', 'Cunning'),
 (datetime.datetime(2013, 2, 5, 18, 34, 25), datetime.datetime(2013, 2, 5, 18, 34, 25), None, 'fd27b86c-7144-11e5-ba71-058fbc01cf0b', 'Turn Your Back', 'Cunning'),
 (datetime.datetime(2013, 2, 5, 18, 34, 39), datetime.datetime(2013, 2, 5, 18, 34, 39), None, 'fd27b86c-7144-11e5-ba71-058fbc01cf0b', 'Cover Your Eyes', 'Cunning')]

In [39]:
%sql SELECT d.user_guid, u.membership_type, d.dog_guid, d.breed \
FROM dogs d, complete_tests c, users u \
WHERE d.breed="Golden Retriever" AND d.dog_guid=c.dog_guid AND d.user_guid=u.user_guid \
GROUP BY d.user_guid, u.membership_type, d.dog_guid, d.breed;

 * mysql://studentuser:***@localhost/dognitiondb
711 rows affected.


[('ce13507c-7144-11e5-ba71-058fbc01cf0b', 1, 'fd27b79a-7144-11e5-ba71-058fbc01cf0b', 'Golden Retriever'),
 ('ce135ab8-7144-11e5-ba71-058fbc01cf0b', 1, 'fd27b6b4-7144-11e5-ba71-058fbc01cf0b', 'Golden Retriever'),
 ('ce135cf2-7144-11e5-ba71-058fbc01cf0b', 2, 'fd7af63a-7144-11e5-ba71-058fbc01cf0b', 'Golden Retriever'),
 ('ce1389d4-7144-11e5-ba71-058fbc01cf0b', 1, 'fd27efb2-7144-11e5-ba71-058fbc01cf0b', 'Golden Retriever'),
 ('ce21f122-7144-11e5-ba71-058fbc01cf0b', 2, 'fd3d03fc-7144-11e5-ba71-058fbc01cf0b', 'Golden Retriever'),
 ('ce220bb2-7144-11e5-ba71-058fbc01cf0b', 2, 'fd3d10cc-7144-11e5-ba71-058fbc01cf0b', 'Golden Retriever'),
 ('ce2237f4-7144-11e5-ba71-058fbc01cf0b', 2, 'fd3d3b24-7144-11e5-ba71-058fbc01cf0b', 'Golden Retriever'),
 ('ce2243e8-7144-11e5-ba71-058fbc01cf0b', 2, 'fd3d4b8c-7144-11e5-ba71-058fbc01cf0b', 'Golden Retriever'),
 ('ce2247a8-7144-11e5-ba71-058fbc01cf0b', 1, 'fd43d9de-7144-11e5-ba71-058fbc01cf0b', 'Golden Retriever'),
 ('ce2258a6-7144-11e5-ba71-058fbc01cf0b', 2, 'fd6487c4-7144-11e5-ba71-058fbc01cf0b', 'Golden Retriever'),
 ('ce22680a-7144-11e5-ba71-058fbc01cf0b', 1, 'fd41e28c-7144-11e5-ba71-058fbc01cf0b', 'Golden Retriever'),
 ('ce240b24-7144-11e5-ba71-058fbc01cf0b', 2, 'fd3fe96e-7144-11e5-ba71-058fbc01cf0b', 'Golden Retriever'),
 ('ce24476a-7144-11e5-ba71-058fbc01cf0b', 2, 'fd404ff8-7144-11e5-ba71-058fbc01cf0b', 'Golden Retriever'),
 ('ce245cdc-7144-11e5-ba71-058fbc01cf0b', 2, 'fd4059e4-7144-11e5-ba71-058fbc01cf0b', 'Golden Retriever'),
 ('ce245cdc-7144-11e5-ba71-058fbc01cf0b', 2, 'fd45eb98-7144-11e5-ba71-058fbc01cf0b', 'Golden Retriever'),
 ('ce246be6-7144-11e5-ba71-058fbc01cf0b', 2, 'fd40738e-7144-11e5-ba71-058fbc01cf0b', 'Golden Retriever'),
 ('ce246be6-7144-11e5-ba71-058fbc01cf0b', 2, 'fd45d6ee-7144-11e5-ba71-058fbc01cf0b', 'Golden Retriever'),
 ('ce249774-7144-11e5-ba71-058fbc01cf0b', 1, 'fd408ed2-7144-11e5-ba71-058fbc01cf0b', 'Golden Retriever'),
 ('ce249c6a-7144-11e5-ba71-058fbc01cf0b', 2, 'fd40945e-7144-11e5-ba71-058fbc01cf0b', 'Golden Retriever'),
 ('ce24a106-7144-11e5-ba71-058fbc01cf0b', 2, 'fd40f822-7144-11e5-ba71-058fbc01cf0b', 'Golden Retriever'),
 ('ce24c1c2-7144-11e5-ba71-058fbc01cf0b', 2, 'fd40ee72-7144-11e5-ba71-058fbc01cf0b', 'Golden Retriever'),
 ('ce24d572-7144-11e5-ba71-058fbc01cf0b', 2, 'fd4107cc-7144-11e5-ba71-058fbc01cf0b', 'Golden Retriever'),
 ('ce24d914-7144-11e5-ba71-058fbc01cf0b', 2, 'fd410cb8-7144-11e5-ba71-058fbc01cf0b', 'Golden Retriever'),
 ('ce24de50-7144-11e5-ba71-058fbc01cf0b', 1, 'fd411690-7144-11e5-ba71-058fbc01cf0b', 'Golden Retriever'),
 ('ce24e116-7144-11e5-ba71-058fbc01cf0b', 1, 'fd4118e8-7144-11e5-ba71-058fbc01cf0b', 'Golden Retriever'),
 ('ce24e684-7144-11e5-ba71-058fbc01cf0b', 2, 'fd411b0e-7144-11e5-ba71-058fbc01cf0b', 'Golden Retriever'),
 ('ce253a6c-7144-11e5-ba71-058fbc01cf0b', 2, 'fd418b02-7144-11e5-ba71-058fbc01cf0b', 'Golden Retriever'),
 ('ce2556a0-7144-11e5-ba71-058fbc01cf0b', 1, 'fd41c5f4-7144-11e5-ba71-058fbc01cf0b', 'Golden Retriever'),
 ('ce255c54-7144-11e5-ba71-058fbc01cf0b', 2, 'fd41c84c-7144-11e5-ba71-058fbc01cf0b', 'Golden Retriever'),
 ('ce257fc2-7144-11e5-ba71-058fbc01cf0b', 2, 'fd41db34-7144-11e5-ba71-058fbc01cf0b', 'Golden Retriever'),
 ('ce2585a8-7144-11e5-ba71-058fbc01cf0b', 2, 'fd41e232-7144-11e5-ba71-058fbc01cf0b', 'Golden Retriever'),
 ('ce258a08-7144-11e5-ba71-058fbc01cf0b', 2, 'fd41e796-7144-11e5-ba71-058fbc01cf0b', 'Golden Retriever'),
 ('ce259dd6-7144-11e5-ba71-058fbc01cf0b', 1, 'fd421400-7144-11e5-ba71-058fbc01cf0b', 'Golden Retriever'),
 ('ce25c306-7144-11e5-ba71-058fbc01cf0b', 1, 'fd423f70-7144-11e5-ba71-058fbc01cf0b', 'Golden Retriever'),
 ('ce25caa4-7144-11e5-ba71-058fbc01cf0b', 2, 'fd4244f2-7144-11e5-ba71-058fbc01cf0b', 'Golden Retriever'),
 ('ce25f4de-7144-11e5-ba71-058fbc01cf0b', 2, 'fd4275a8-7144-11e5-ba71-058fbc01cf0b', 'Golden Retriever'),
 ('ce26092e-7144-11e5-ba71-058fbc01cf0b', 1, 'fd42933a-7144-11e5-ba71-058fbc01cf0b', 'Golden Retriever'),
 ('ce261a40-7144-11e5-ba71-058fbc01cf0b', 1, 'fd42a29e-7144-11e5-ba71-058fbc01

**질문 7: Dognition 데이터베이스에 North Carolina에 사는 고유의 Golden Retrievers들은 몇마리가 있나요 (정답은 30 마리입니다)?**

In [45]:
%sql SELECT d.user_guid, u.membership_type, d.dog_guid, d.breed,u.state \
FROM dogs d, complete_tests c, users u \
WHERE d.breed="Golden Retriever" AND u.state="NC" AND d.dog_guid=c.dog_guid AND d.user_guid=u.user_guid \
GROUP BY d.user_guid, u.membership_type, d.dog_guid, d.breed, u.state;

 * mysql://studentuser:***@localhost/dognitiondb
29 rows affected.


[('ce135cf2-7144-11e5-ba71-058fbc01cf0b', 2, 'fd7af63a-7144-11e5-ba71-058fbc01cf0b', 'Golden Retriever', 'NC'),
 ('ce2247a8-7144-11e5-ba71-058fbc01cf0b', 1, 'fd43d9de-7144-11e5-ba71-058fbc01cf0b', 'Golden Retriever', 'NC'),
 ('ce2258a6-7144-11e5-ba71-058fbc01cf0b', 2, 'fd6487c4-7144-11e5-ba71-058fbc01cf0b', 'Golden Retriever', 'NC'),
 ('ce24a106-7144-11e5-ba71-058fbc01cf0b', 2, 'fd40f822-7144-11e5-ba71-058fbc01cf0b', 'Golden Retriever', 'NC'),
 ('ce24e684-7144-11e5-ba71-058fbc01cf0b', 2, 'fd411b0e-7144-11e5-ba71-058fbc01cf0b', 'Golden Retriever', 'NC'),
 ('ce253a6c-7144-11e5-ba71-058fbc01cf0b', 2, 'fd418b02-7144-11e5-ba71-058fbc01cf0b', 'Golden Retriever', 'NC'),
 ('ce2758c4-7144-11e5-ba71-058fbc01cf0b', 1, 'fd44236c-7144-11e5-ba71-058fbc01cf0b', 'Golden Retriever', 'NC'),
 ('ce339990-7144-11e5-ba71-058fbc01cf0b', 1, 'fd5b57a8-7144-11e5-ba71-058fbc01cf0b', 'Golden Retriever', 'NC'),
 ('ce355c1c-7144-11e5-ba71-058fbc01cf0b', 1, 'fd54622c-7144-11e5-ba71-058fbc01cf0b', 'Golden Retriever', 'NC'),
 ('ce3c3f96-7144-11e5-ba71-058fbc01cf0b', 1, 'fd5b35e8-7144-11e5-ba71-058fbc01cf0b', 'Golden Retriever', 'NC'),
 ('ce3d35d6-7144-11e5-ba71-058fbc01cf0b', 1, 'fd5c1c6a-7144-11e5-ba71-058fbc01cf0b', 'Golden Retriever', 'NC'),
 ('ce714f56-7144-11e5-ba71-058fbc01cf0b', 1, 'fd702714-7144-11e5-ba71-058fbc01cf0b', 'Golden Retriever', 'NC'),
 ('ce714f56-7144-11e5-ba71-058fbc01cf0b', 1, 'fd7032f4-7144-11e5-ba71-058fbc01cf0b', 'Golden Retriever', 'NC'),
 ('ce72e33e-7144-11e5-ba71-058fbc01cf0b', 3, 'fd72138a-7144-11e5-ba71-058fbc01cf0b', 'Golden Retriever', 'NC'),
 ('ce72e33e-7144-11e5-ba71-058fbc01cf0b', 3, 'fd72168c-7144-11e5-ba71-058fbc01cf0b', 'Golden Retriever', 'NC'),
 ('ce76f7e4-7144-11e5-ba71-058fbc01cf0b', 2, 'fd76c970-7144-11e5-ba71-058fbc01cf0b', 'Golden Retriever', 'NC'),
 ('ce76f7e4-7144-11e5-ba71-058fbc01cf0b', 2, 'fd76c9d4-7144-11e5-ba71-058fbc01cf0b', 'Golden Retriever', 'NC'),
 ('ce7817be-7144-11e5-ba71-058fbc01cf0b', 2, 'fd77e166-7144-11e5-ba71-058fbc01cf0b', 'Golden Retriever', 'NC'),
 ('ce79f4da-7144-11e5-ba71-058fbc01cf0b', 1, 'fd7a2552-7144-11e5-ba71-058fbc01cf0b', 'Golden Retriever', 'NC'),
 ('ce7a7428-7144-11e5-ba71-058fbc01cf0b', 2, 'fd7aaaae-7144-11e5-ba71-058fbc01cf0b', 'Golden Retriever', 'NC'),
 ('ce7a7428-7144-11e5-ba71-058fbc01cf0b', 2, 'fd7aab1c-7144-11e5-ba71-058fbc01cf0b', 'Golden Retriever', 'NC'),
 ('ce7b943e-7144-11e5-ba71-058fbc01cf0b', 1, 'fd7c3ca2-7144-11e5-ba71-058fbc01cf0b', 'Golden Retriever', 'NC'),
 ('ce7bd232-7144-11e5-ba71-058fbc01cf0b', 1, 'fd7cc4a6-7144-11e5-ba71-058fbc01cf0b', 'Golden Retriever', 'NC'),
 ('ce7cdc40-7144-11e5-ba71-058fbc01cf0b', 1, 'fd7f4b9a-7144-11e5-ba71-058fbc01cf0b', 'Golden Retriever', 'NC'),
 ('ce7e552a-7144-11e5-ba71-058fbc01cf0b', 1, 'fd82d86e-7144-11e5-ba71-058fbc01cf0b', 'Golden Retriever', 'NC'),
 ('ce825008-7144-11e5-ba71-058fbc01cf0b', 1, 'fdabe5b0-7144-11e5-ba71-058fbc01cf0b', 'Golden Retriever', 'NC'),
 ('ce876d54-7144-11e5-ba71-058fbc01cf0b', 1, 'fdb0b69e-7144-11e5-ba71-058fbc01cf0b', 'Golden Retriever', 'NC'),
 ('ce9390de-7144-11e5-ba71-058fbc01cf0b', 1, 'fdb88702-7144-11e5-ba71-058fbc01cf0b', 'Golden Retriever', 'NC'),
 ('ce96c9a2-7144-11e5-ba71-058fbc01cf0b', 1, 'fdbc31f4-7144-11e5-ba71-058fbc01cf0b', 'Golden Retriever', 'NC')]

** 질문 8: 각 membership type에 review를 남긴 고유의 고객 수는 몇인가요? (NULL 값이 있는 ratings열을 포함하지 않으면 가장 많은 membership type에는 2900개의 고객이 있고, 가장 적은 membership type에는 15명의 고객이 있습니다) **

In [55]:
%sql SELECT COUNT(*) FROM reviews WHERE rating IS NULL;

 * mysql://studentuser:***@localhost/dognitiondb
1 rows affected.


[(2451,)]

In [56]:
%sql SELECT COUNT(*) FROM reviews WHERE rating="None";

 * mysql://studentuser:***@localhost/dognitiondb
1 rows affected.


[(10485,)]

In [61]:
%sql SELECT u.membership_type, COUNT(DISTINCT(r.user_guid)) \
FROM users u, reviews r WHERE r.rating IS NOT NULL AND  u.user_guid=r.user_guid GROUP BY u.membership_type;

 * mysql://studentuser:***@localhost/dognitiondb
5 rows affected.


[(1, 2900), (2, 1120), (3, 238), (4, 816), (5, 15)]

**질문 9: 가장 site_activity 데이터가 많은 3개의 개 품종 (breeds)는 무엇인가요? 이 정보는 script_detail_id가 NULL값이 아닌 걸로 알 수 있습니다 (답은 "Mixed", "Labrador Retriever"와 "Labrador Retriever-Golden Retriever Mix"입니다)**

In [64]:
%sql SELECT DISTINCT(step_type) FROM exam_answers;

 * mysql://studentuser:***@localhost/dognitiondb
7 rows affected.


[('question',),
 ('stopwatch',),
 ('if_question',),
 ('bark',),
 ('timer',),
 ('button',),
 (None,)]